src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/DictionaryLayoutNode.cs

```diff
@@ -55,20 +55,138 @@ private void Validate()
             get;
         }
 
-        public abstract ICollection<NativeLayoutVertexNode> GetTemplateEntries(NodeFactory factory);
+        public TypeSystemEntity OwningMethodOrType => _owningMethodOrType;
 
-        public abstract void EmitDictionaryData(ref ObjectDataBuilder builder, NodeFactory factory, GenericDictionaryNode dictionary);
+        /// <summary>
+        /// Gets a value indicating whether the slot assignment is determined at the node creation time.
+        /// </summary>
+        public abstract bool HasFixedSlots
+        {
+            get;
+        }
+
+        public virtual ICollection<NativeLayoutVertexNode> GetTemplateEntries(NodeFactory factory)
+        {
+            ArrayBuilder<NativeLayoutVertexNode> templateEntries = new ArrayBuilder<NativeLayoutVertexNode>();
+            foreach (var entry in Entries)
+            {
+                templateEntries.Add(entry.TemplateDictionaryNode(factory));
+            }
```

Comment:
Looking at repeated patterns where we transform the Entries enumerable into a series of different collections of exactly the same size as the Entries enumerable makes me think that would should be using ICollection<GenericLookupResult> instead of IEnumerable<GenericLookupResult>. (This would allow us to pre-allocate the right amount of space in the new collection instead of using the standard growth factors.)